serverless-functions/src/functions/features/callback-and-voicemail/studio/wait-experience.protected.js

```diff
@@ -0,0 +1,340 @@
+ *
+ * Much of the logic is the same for callback or voicemail, so we're using this single function to handle both.
+ *
+ * @param {*} context
+ * @param {*} isVoicemail
+ * @param {*} callSid
+ * @param {*} taskSid
+ * @returns
+ */
+async function handleCallbackOrVoicemailSelected(
+  context,
+  isVoicemail,
+  callSid,
+  taskSid
+) {
+  const domain = `https://${context.DOMAIN_NAME}`;
+  const twiml = new Twilio.twiml.VoiceResponse();
+
+  const cancelReason = isVoicemail
+    ? "Opted to leave a voicemail"
+    : "Opted to request a callback";
+  const mode = isVoicemail ? "record-voicemail" : "submit-callback";
+
+  const task = await fetchTask(context, taskSid);
+
+  // Redirect Call to callback or voicemail logic. We need to update the call with a new TwiML URL vs using twiml.redirect() - since
+  // we are still in the waitUrl TwiML execution - and it's not possible to use the <Record> verb in here. We piggyback on the same approach for callbacks,
+  // though technically these could be handled entirely in the waitUrl TwiML
+  const redirectUrl = `${domain}/features/callback-and-voicemail/studio/wait-experience?mode=${mode}&CallSid=${callSid}&enqueuedTaskSid=${taskSid}`;
+  const result = await VoiceOperations.updateCall({
+    context,
+    callSid,
+    params: { method: "POST", url: redirectUrl },
+    attempts: 0,
+  });
+  const { success, status } = result;
+
+  if (success) {
+    //  Cancel (update) the task with handy attributes for reporting
+    await cancelTask(context, task, cancelReason);
+  } else {
+    console.error(
+      `Failed to update call ${callSid} with new TwiML. Status: ${status}`
+    );
+    twiml.say(
+      sayOptions,
+      "Sorry, we were unable to perform this operation. Please remain on the line."
+    );
+    twiml.redirect(
+      `${domain}/features/callback-and-voicemail/studio/wait-experience?mode=main-wait-loop&CallSid=${callSid}&enqueuedTaskSid=${taskSid}&skipGreeting=true`
+    );
+    return twiml;
+  }
+  return "";
+}
+
+exports.handler = async function (context, event, callback) {
+  const domain = `https://${context.DOMAIN_NAME}`;
+  const twiml = new Twilio.twiml.VoiceResponse();
+
+  // Retrieve options
+  const { sayOptions, holdMusicUrl } = options;
+
+  const {
+    Digits,
+    CallSid,
+    enqueuedWorkflowSid,
+    mode,
+    enqueuedTaskSid,
+    skipGreeting,
+  } = event;
+
+  let message = "";
+
+  switch (mode) {
+    case "initialize":
+      // Initial logic to find the associated task for the call, and propagate it through to the rest of the TwiML execution
+      // If the lookup fails to find the task, the remaining TwiML logic will not offer any callback or voicemail options.
+      const enqueuedTask = await getPendingTaskByCallSid(
+        context,
+        CallSid,
+        enqueuedWorkflowSid
+      );
+
+      const redirectUrlNoTaskFound = `${domain}/features/callback-and-voicemail/studio/wait-experience?mode=main-wait-loop&CallSid=${CallSid}`;
```

Comment:
Looks like this var is used also in the case below when a task _is_ found, which is a bit confusing, could it be named more generically like `redirectUrl`?

Reply:
Done